### PR TITLE
LibVideo: Add the concept of codec identifiers

### PR DESCRIPTION
**Pull request by feliwir**
This is required for detecting which decoder should be used. Only a small subset of codecs identifiers is added for now

### CODE REVIEW

File: Userland/Libraries/LibVideo/CodecID.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    VP9,
+    // MPEG
+    H261,
+    H262,
+    H263,
+    H264,
+    H265,
+    MPEG1,
+    MPEG2 = H262,
+    MPEG4 = H263,
```
Comment (skyrising) [inline Userland/Libraries/LibVideo/CodecID.h:30]: Is this a typo that should be H264?
MPEG4 isn't the name of a codec anyway (only MPEG-4 Part 10), so I'd suggest this change:
```suggestion
    AVC = H264,
    HEVC = H265,
```

File: Userland/Libraries/LibVideo/CodecID.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#include <AK/Variant.h>
+
+namespace Video {
+
+enum class CodecID : u32 {
+    Unknown,
+    // On2 / Google
+    VP8,
+    VP9,
+    // MPEG
```
Comment (skyrising) [inline Userland/Libraries/LibVideo/CodecID.h:22]: This comment is technically incorrect, maybe make it two comments for `// ITU-T` and `// MPEG`

File: Userland/Libraries/LibVideo/Containers/Matroska/MatroskaDemuxer.cpp [hunk 3]
```diff
@@ -71,6 +71,35 @@ DecoderErrorOr<MatroskaDemuxer::TrackStatus*> MatroskaDemuxer::get_track_status(
...
 
+CodecID MatroskaDemuxer::get_codec_id_for_string(DeprecatedFlyString const& codec_id)
+{
+    dbgln_if(MATROSKA_DEBUG, "Codec ID: {}", codec_id);
+    // On2 / Google codecs
+    if (codec_id == "V_VP8")
+        return CodecID::VP8;
+    if (codec_id == "V_VP9")
+        return CodecID::VP9;
+    // MPEG codecs
```
Comment (skyrising) [inline Userland/Libraries/LibVideo/Containers/Matroska/MatroskaDemuxer.cpp:82]: Similar to above, this is a mix of ITU-T and MPEG names, maybe stick to the MPEG aliases, since those are the ones used in MKV's naming scheme.
Also you're missing a comment for AOMedia. But I think here we can just leave the comments out entirely.

File: Userland/Libraries/LibVideo/PlaybackManager.cpp [hunk 4]
```diff
@@ -52,9 +52,19 @@ DecoderErrorOr<NonnullOwnPtr<PlaybackManager>> PlaybackManager::create(NonnullOw
...
 
-    auto decoder = DECODER_TRY_ALLOC(try_make<VP9::Decoder>());
+    OwnPtr<VideoDecoder> decoder;
+    switch (track.codec_id()) {
+    case CodecID::VP9:
+        decoder = DECODER_TRY_ALLOC(try_make<VP9::Decoder>());
+        break;
+
+    default:
+        dbgln("Unsupported codec: {}", (int)track.codec_id());
```
Comment (skyrising) [inline Userland/Libraries/LibVideo/PlaybackManager.cpp:62]:
```suggestion
        dbgln("Unsupported codec: {}", to_underlying(track.codec_id()));
```

But a formatter would be even better.

File: Userland/Libraries/LibVideo/CodecID.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    H261,
+    MPEG1,
+    H262,
+    MPEG2 = H262,
+    H263,
+    AVC = H263,
```
Comment (skyrising) [inline Userland/Libraries/LibVideo/CodecID.h:24]: Still wrong, AVC = H264

File: Userland/Libraries/LibVideo/CodecID.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            break;
+        case Video::CodecID::H263:
+            codec = "H263"sv;
+            break;
+        case Video::CodecID::H264:
+            codec = "H264"sv;
+            break;
+        case Video::CodecID::H265:
+            codec = "H265"sv;
+            break;
```
Comment (skyrising) [inline Userland/Libraries/LibVideo/CodecID.h:68]: These could include the `.`, but I guess it's a question of whether we want the formatter to print the human-readable name in general or specifically the name of the enum constant.

File: Userland/Libraries/LibVideo/CodecID.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    VP8,
+    VP9,
+    // MPEG
+    H261,
+    MPEG1,
+    H262,
+    MPEG2 = H262,
+    H263,
+    H264,
+    AVC = H264,
```
Comment (Zaggy1024) [inline Userland/Libraries/LibVideo/CodecID.h:25]: I'm not sure if we benefit at all from aliasing these, it seems like that could just lead to confusion. Better to just have a consistent naming style and use it everywhere we can.

File: Userland/Libraries/LibVideo/CodecID.h (hunk 6 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibVideo/CodecID.h:68]: I would agree that it should include the `.` as well.

File: Userland/Libraries/LibVideo/Containers/Matroska/Reader.cpp [hunk 8]
```diff
@@ -985,7 +985,8 @@ ErrorOr<DeprecatedString> Streamer::read_string()
     auto string_length = TRY(read_variable_size_integer());
     if (remaining() < string_length)
         return Error::from_string_literal("String length extends past the end of the stream");
-    auto string_value = DeprecatedString(data_as_chars(), string_length);
+    auto string_data = data_as_chars();
+    auto string_value = DeprecatedString(string_data, strnlen(string_data, string_length));
```
Comment (Zaggy1024) [inline Userland/Libraries/LibVideo/Containers/Matroska/Reader.cpp:989]: This should probably go in a separate commit, since it was never correct to begin with.

File: Userland/Libraries/LibVideo/Containers/Matroska/MatroskaDemuxer.cpp [hunk 9]
```diff
@@ -43,7 +43,7 @@ DecoderErrorOr<Vector<Track>> MatroskaDemuxer::get_tracks_for_type(TrackType typ
     Vector<Track> tracks;
     TRY(m_reader.for_each_track_of_type(matroska_track_type, [&](TrackEntry const& track_entry) -> DecoderErrorOr<IterationDecision> {
         VERIFY(track_entry.track_type() == matroska_track_type);
-        Track track(type, track_entry.track_number());
+        Track track(type, track_entry.track_number(), get_codec_id_for_string(track_entry.codec_id()));
```
Comment (Zaggy1024) [inline Userland/Libraries/LibVideo/Containers/Matroska/MatroskaDemuxer.cpp:46]: Let's make a `Demuxer::get_codec_for_track(Track track)` function so this can be a lazy getter instead of having to set the codec ID for tracks we may not use. Here, it's convenient since we already read out the string, but other containers may not want to do that ahead of time.

Later on, `get_codec_for_track()` may be used to request more track metadata if there is anything that we usually need around the same time that we need to get the track's codec (perhaps `TrackType`-specific information like nominal resolution).

File: Userland/Libraries/LibVideo/PlaybackManager.cpp [hunk 10]
```diff
@@ -52,9 +52,19 @@ DecoderErrorOr<NonnullOwnPtr<PlaybackManager>> PlaybackManager::create(NonnullOw
...
-    auto decoder = DECODER_TRY_ALLOC(try_make<VP9::Decoder>());
+    OwnPtr<VideoDecoder> decoder;
+    switch (track.codec_id()) {
+    case CodecID::VP9:
+        decoder = DECODER_TRY_ALLOC(try_make<VP9::Decoder>());
+        break;
+
+    default:
+        dbgln("Unsupported codec: {}", track.codec_id());
+        return DecoderError::with_description(DecoderErrorCategory::Invalid, "Unsupported codec"sv);
```
Comment (Zaggy1024) [inline Userland/Libraries/LibVideo/PlaybackManager.cpp:63]: We can format the codec ID into a string for `DecoderError` here to be more helpful to users here.

File: Userland/Libraries/LibVideo/VP9/Decoder.h [hunk 11]
```diff
@@ -29,6 +30,7 @@ class Decoder : public VideoDecoder {
     ~Decoder() override { }
     /* (8.1) General */
     DecoderErrorOr<void> receive_sample(ReadonlyBytes) override;
+    static bool is_codec_supported(CodecID codec) { return codec == CodecID::VP9; }
```
Comment (Zaggy1024) [inline Userland/Libraries/LibVideo/VP9/Decoder.h:33]: This is unused currently, so we should omit it for now. We may have a use for it in the future, but we can just add it then.

**Review by Zaggy1024 — COMMENTED** [review]
Mostly some small nitpicks, but thanks for doing this!

File: Userland/Libraries/LibVideo/CodecID.h (hunk 6 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibVideo/CodecID.h:68]: > These could include the `.`, but I guess it's a question of whether we want the formatter to print the human-readable name in general or specifically the name of the enum constant.

Tried to reply to this and made a separate review comment instead, whoops. The formatter should definitely use human-friendly names for these, so it should include the ".", if we're going to use the ITU-T names.

Comment (skyrising) [reply Userland/Libraries/LibVideo/CodecID.h:68]: H.264

File: Userland/Libraries/LibVideo/CodecID.h (hunk 7 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibVideo/CodecID.h:25]: Given that this is an internal enum, and it doesn't differentiate between these aliases, it doesn't do us any good to use multiple names. If we were to keep the naming style from the container, perhaps there's an argument for that, but then if we display the codec to the user eventually, they may not know that H.265 is the same as HEVC.

Anyone writing code to make use of these values should know what the different names mean. When displaying the codec to the user, we should either stick with one name, or give all aliases at once in the human-friendly text.

**Comment by Zaggy1024** [comment]
LGTM!
